src/vmm/src/vstate/memory.rs

```diff
@@ -53,12 +53,39 @@ pub enum MemoryError {
 }
 
 /// Defines the interface for snapshotting memory.
-pub trait SnapshotMemory
+pub trait GuestMemoryExtension
```

Comment:
I don't quite like the name `Extension`. It looks like we are adding all the stuff we are missing from `vm-memory::GuestMemory` here (`with_size()`, `from_raw_regions()`, `from_raw_regions_unguarded()`). Would it make sense to upstream it? @roypat wdyt?
There is still a method `from_state()` that refers back to the original name `SnapshotMemory`. This looks like a separate trait that should stay in FC as rust-vmm's vm-memory doesn't have a clue about the state.

Reply:
Mh, I think the decision around these was that vm-memory has no interest in taking on firecracker's concept of "guard pages".

Reply:
I also don't think we should upstream our FC specific memory creation process.
Regarding the `GuestMemoryExtension` name, the original idea was that this trait just 'extends' what `GuestMemoryMmap` already has. But I don't mind changing it to smth better sounding. Any ideas?

Reply:
What about the following:
- It looks like the purpose of guarding the guest memory has been lost over time. I think we should either document why we are doing that or remove it (and we won't need to think how to test it)
- I agree, the creation thing may be opinionated to FC (so don't upstream it), but then let's create a trait only for those methods with a meaningful name (`factory` or something like)
- We have a family of `from_raw_regions_*` methods which are either used internally by the implementation or in testing. Is it possible to remove the helpers from the trait as standalone private functions and extract the rest into a trait that is used specifically for testing? Also, does `raw` part of the name add any meaning? Can we drop it?
- There is a family of snapshot-related methods (`from_state`, `describe`, `make_dirty`, `dump`, `dump_dirty`). Let's extract them into a trait that is specific to snapshot. Plus maybe also include `get_dirty_bitmap`, which seems to partially belong here as well (I know it might not be worth it).

I hope thus we'll have a bit more structure than a toolbox trait.
Would that make sense?

Reply:
> It looks like the purpose of guarding the guest memory has been lost over time. I think we should either document why we are doing that or remove it (and we won't need to think how to test it)

I will look into it

> I agree, the creation thing may be opinionated to FC (so don't upstream it), but then let's create a trait only for those methods with a meaningful name (factory or something like)

Please, no factories. We are not a manufacturing site. 

> We have a family of from_raw_regions_* methods which are either used internally by the implementation or in testing. Is it possible to remove the helpers from the trait as standalone private functions and extract the rest into a trait that is used specifically for testing? Also, does raw part of the name add any meaning? Can we drop it?
There is a family of snapshot-related methods (from_state, describe, make_dirty, dump, dump_dirty). Let's extract them into a trait that is specific to snapshot. Plus maybe also include get_dirty_bitmap, which seems to partially belong here as well (I know it might not be worth it).

The idea here is to have them in one trait for ease of use. It does not really matter if they are public or not, everything is used internally only anyway. The `raw` part is just to differentiate from `from_regions` method that `GuestMemoryMmap` already has.
The only alternative I see is to just remove trait and have a bunch of free standing functions. Personally I don't like this approach, but it is better than having multiple traits.